cmd/kubeadm/app/util/system/logutils.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package system
+
+// LogMessage is a struct for logs
+type LogMessage struct {
```

Comment:
I'm not sure we should adopt such thing. we should return array of strings for info, an array or one warn, and an err in case of error, the caller should decide of the verbosity to use

Reply:
I don't think so. there are 3 different things we return.
Log messages that can be info/warn/error ect
None critical Errors
Critical Errors,
I originally thought just strings but it would leave a lot of info out. and we would no longer pass the errors out anymore, just some strings.

Reply:
I still think we should aggregate errs and logs and let the caller output them cc @neolit123 @rosti

Reply:
Ok. I can do whatever you all think is best! If that is what you all want it's what I will do! If you read my own comments I am not a fan of this entire file. I just did not know of a way without just returning a flat string. and I thought that the caller lost a lot of info with a method like that. I would also be up for a zoom chat if you all wanted.

Reply:
I'd like to hear from other folks before we make any decision

Reply:
:+1: this is why I posted it in [WIP] as I stated I don't like it how it is and wanted feedback!  Thanks for the comments! I'm just letting you know why I did not initially go with the `[]string` method :)

Reply:
@Klaven @yastij 
we should have probably discussed this a little or clarified in the related ticket.
we need a logging shim and we should not define custom objects and additional functional returns.

the wheel that was already invented here is called logr:
https://github.com/go-logr/logr/blob/master/logr.go

and klogr:
https://github.com/kubernetes/klog/blob/master/klogr/klogr.go

one way of approaching the problem is:

- create new klogr instances on the side of the system validator callers using `New()`.
- pass these instances as `logr.Logger` to the validators methods and constructors.
- use `mylog.Info` on the side of the validators.

but this forces the users of the validators to implement logr and logr is opinionated and complicated.

our alternative way here (that i prefer) is to just a pass logging shim from the validator callers to the validators:

```
// this is defined on the side of the validators
// we don't need more than Info* for now
type Logger interface {
  Info(args ...interface{})
  Infof(format string, args ...interface{})
}
```

so on the side of the callers one would define:
```
type myLogger struct {}

// the caller chooses the verbosity level, in this case 1
func (myLogger) Info(args ...interface{}) {
  klog.V(1).Info(args...)
}

func (myLogger) Infof(format string, args ...interface{}) {
  klog.V(1).Infof(format, args...)
}

...
myLog := &myLogger{}
// and pass this this to the validators as `system.Logger`
```

alternatively instead of passing this instance to all validator constructors etc.

the `system` package can have:
```
var logger Logger
var mutex sync.Mutex

func SetLogger(log Logger) {
  mutex.Lock()
  defer mutex.Unlock()
  logger = log
}
```

and then the validator code can call `logger.Infof`, which should panic if the consumer of the system validators has not called `SetLogger(...)`

Reply:
I'm fine with this, as long as we keep the interface in sync (although I don't expect it to change that much over time)

Reply:
@neolit123 I like the shim, originally when I thought of the shim it sounded like a lot of reinventing, aka, `info`, `warn` and `error` as well as the ability to set verbosity. meaning everyone that tried to call these function would have to go though a bit of work to implement. But if I can just assume that all things are the same (and currently are NOT, we log with `info` and `error`) then this is in fact simple. I don't think the `SetLogger` is awesome just because it could be frustrating trying to use a function that requires some state set before hand that might not be self evident.

Reply:
I also like `logr` :thinking: and probably would have gone down this path last night had I seen it. would it be possible for me to create a `klog` `logr` shim, so if you wanted to use `klog` it was already created and easy? and then it would just be anyone that wanted to log without klog that has to go though the painful stuff? I feel like just logging everything as `V(1).Info` sucks a little.

Reply:
oh. klogr already is that shim :) sorry. missed that point. is there a reason you don't like this solution? anyone not using `klog` would need to implement it but is that a big problem?